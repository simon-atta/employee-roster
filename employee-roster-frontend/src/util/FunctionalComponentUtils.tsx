/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List } from 'immutable';
import { useEffect, useRef, useState } from 'react';
import { useHistory } from 'react-router';
import { Sorter } from 'types';
import { DataTableUrlProps } from 'ui/components/DataTable';

export function useUrlState(urlProperty: string, initialValue?: string):
[string|null, (newValue: string|null) => void] {
  const history = useHistory();
  const searchParams = new URLSearchParams(history.location.search);
  const [state, setState] = useState(searchParams.get(urlProperty) || initialValue || null);
  return [
    state || null,
    (newValue) => {
      setState(newValue);
      const newSearchParams = new URLSearchParams(history.location.search);
      if (newValue !== null) {
        newSearchParams.set(urlProperty, newValue);
      } else {
        newSearchParams.delete(urlProperty);
      }
      history.push(`${history.location.pathname}?${newSearchParams.toString()}`);
    }];
}

// From https://overreacted.io/making-setinterval-declarative-with-react-hooks/
export function useInterval(callback: Function, delay: number|null) {
  const savedCallback = useRef<Function>();

  // Remember the latest callback.
  useEffect(() => {
    savedCallback.current = callback;
  }, [callback]);

  // Set up the interval.
  // eslint-disable-next-line consistent-return
  useEffect(() => {
    function tick() {
      (savedCallback.current as Function)();
    }
    if (delay !== null) {
      const id = setInterval(tick, delay);
      return () => clearInterval(id);
    }
  }, [delay]);
}


export interface PaginationData<T> {
  filterText: string;
  page: number;
  itemsPerPage: number;
  filteredRows: List<T>;
  numOfFilteredRows: number;
  rowsInPage: List<T>;
  isReversed: boolean;
}

export function usePageableData<T>(urlProps: Omit<DataTableUrlProps, 'sortBy'>, data: T[],
  valueToText: (value: T) => string[], sortBy: Sorter<T>): PaginationData<T> {
  const filterText = urlProps.filter || '';
  const page = parseInt(urlProps.page || '1', 10);
  const itemsPerPage = parseInt(urlProps.itemsPerPage || '10', 10);
  const isReversed = urlProps.asc === 'false';

  const sortedRows = List.of(...data)
    .sort(isReversed ? (a, b) => sortBy(b, a) : sortBy);

  const filteredRows = sortedRows
    .filter(value => valueToText(value)
      .filter(text => text.toLowerCase().indexOf(filterText.toLowerCase()) !== -1).length > 0);

  const numOfFilteredRows = filteredRows.size;

  const rowsInPage = filteredRows
    .skip((page - 1) * itemsPerPage)
    .take(itemsPerPage);

  return {
    filterText,
    page,
    itemsPerPage,
    filteredRows,
    numOfFilteredRows,
    rowsInPage,
    isReversed,
  };
}
