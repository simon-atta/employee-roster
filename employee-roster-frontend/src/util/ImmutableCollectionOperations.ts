/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DomainObject } from 'domain/DomainObject';
import DomainObjectView from 'domain/DomainObjectView';
import { Map, Seq } from 'immutable';

export function toggleElement<T>(collection: T[], element: T): T[] {
  if (collection.indexOf(element) !== -1) {
    return collection.filter(e => e !== element);
  }

  return [...collection, element];
}

interface ObjectWithKeys {
  [key: string]: any;
}

export function objectWithout<F>(obj: F, ...without: (keyof F)[]): F {
  const copy: ObjectWithKeys = {
    ...obj,
  };
  without.forEach(key => delete copy[key as string]);
  return copy as F;
}

export function mapDomainObjectToView<T extends DomainObject>(obj: T|DomainObjectView<T>): DomainObjectView<T> {
  const result = {} as ObjectWithKeys;
  const objWithKeys = obj as ObjectWithKeys;
  Object.keys(objWithKeys).forEach((key) => {
    if (objWithKeys[key] !== null && objWithKeys[key].id !== undefined) {
      result[key] = objWithKeys[key].id;
    } else if (Array.isArray(objWithKeys[key]) && objWithKeys[key].length > 0
        && objWithKeys[key].find((item: any) => item && item.id !== undefined) !== undefined) {
      result[key] = objWithKeys[key].map((ele: DomainObject) => ele.id);
    } else {
      result[key] = objWithKeys[key];
    }
  });
  return result as DomainObjectView<T>;
}

export function createIdMapFromList<T extends DomainObject>(collection: T[]): Map<number, DomainObjectView<T>> {
  let map = Map<number, DomainObjectView<T>>();
  collection.forEach((ele) => { (map = map.set(ele.id as number, mapDomainObjectToView(ele))); });
  return map;
}

export interface FluentValue<V, M> {
  result: V;
  then: (map: M) => FluentValue<V, M>;
}
export function conditionally<K, V>(seq: Seq<K, V>, mapper: (seq: Seq<K, V>) => Seq<K, V> | undefined):
FluentValue<Seq<K, V>, (seq: Seq<K, V>) => Seq<K, V> | undefined> {
  const out = mapper(seq) || seq;
  return {
    result: out,
    then: newMapper => conditionally(out, newMapper),
  };
}
