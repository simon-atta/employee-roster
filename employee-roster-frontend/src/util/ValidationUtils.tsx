/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react';
import { Predicate } from 'types';
import { InputGroupText, Tooltip, List, ListItem } from '@patternfly/react-core';
import { WarningTriangleIcon } from '@patternfly/react-icons';

export interface Validator<T> {
  predicate: Predicate<T>;
  errorMsg: (value: T) => string;
}
export type ValidatorGroupMap<K extends string, T> = Record<K, Validator<T>>;
export interface ValidationResult<K> {
  isValid: boolean;
  validationErrors: K[];
  showValidationErrors: (...validatorGroups: K[]) => React.ReactNode;
}

export interface ValidationListResult<K> {
  isValid: boolean;
  validationErrors: K[][];
  showValidationErrors: (index: number, ...validatorGroups: K[]) => React.ReactNode;
}


export function useValidators<K extends string, T>(value: T, validators: ValidatorGroupMap<K, T>): ValidationResult<K> {
  const validationErrors: K[] = Object.keys(validators).filter((key) => {
    const validator = validators[key as K];
    return !validator.predicate(value);
  }) as K[];
  const isValid = validationErrors.length === 0;
  const showValidationErrors = (...shownValidatorGroups: K[]) => {
    const invalidatedShownValidatorGroups = shownValidatorGroups.filter(group => validationErrors.includes(group));
    if (invalidatedShownValidatorGroups.length === 0) {
      // If there are no validation errors within the specified validation groups,
      // then nothing is rendered
      return null;
    }

    return (
      <InputGroupText key={0}>
        <Tooltip
          content={(
            <List
              style={{
                color: 'white',
              }}
            >
              {invalidatedShownValidatorGroups.map(key => (
                <ListItem key={key}>
                  {validators[key].errorMsg(value)}
                </ListItem>
              ))
              }
            </List>
          )}
        >
          <WarningTriangleIcon />
        </Tooltip>
      </InputGroupText>
    );
  };
  return {
    isValid,
    validationErrors,
    showValidationErrors,
  };
}

export function useListValidators<K extends string, T>(value: T[], validators: ValidatorGroupMap<K, T>):
ValidationListResult<K> {
  const validationErrors: K[][] = value.map(item => Object.keys(validators).filter((key) => {
    const validator = validators[key as K];
    return !validator.predicate(item);
  }) as K[]);
  const isValid = validationErrors.reduce((prev, curr) => prev && curr.length === 0, true);
  const showValidationErrors = (index: number, ...shownValidatorGroups: K[]) => {
    const invalidatedShownValidatorGroups = shownValidatorGroups.filter(group => validationErrors[index]
      .includes(group));
    if (invalidatedShownValidatorGroups.length === 0) {
      // null is rendered as null
      return null;
    }

    return (
      <InputGroupText key={0}>
        <Tooltip
          content={(
            <List
              style={{
                color: 'white',
              }}
            >
              {invalidatedShownValidatorGroups.map(key => (
                <ListItem key={key}>
                  {validators[key].errorMsg(value[index])}
                </ListItem>
              ))
              }
            </List>
          )}
        >
          <WarningTriangleIcon />
        </Tooltip>
      </InputGroupText>
    );
  };
  return {
    isValid,
    validationErrors,
    showValidationErrors,
  };
}
