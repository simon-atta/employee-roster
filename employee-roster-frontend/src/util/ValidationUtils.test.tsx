/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { useValidators, Validator, useListValidators } from 'util/ValidationUtils';

describe('Validation Utils', () => {
  const lengthGreaterThan2: Validator<string> = {
    predicate: str => str.length > 2,
    errorMsg: str => `"${str}" should have a length of at least 3`,
  };
  const startsWithA: Validator<string> = {
    predicate: str => str.length === 0 || str.charAt(0) === 'A',
    errorMsg: str => `"${str}" should start with "A"`,
  };

  it('isValid should be true iff all validators predicates return true', () => {
    let validationResult = useValidators('An', {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(validationResult.isValid).toEqual(false);
    expect(validationResult.validationErrors).toEqual(['lengthGreaterThan2']);
    expect(
      validationResult.showValidationErrors('startsWithA'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors('lengthGreaterThan2'),
    ).not.toEqual(null);
    expect(
      validationResult.showValidationErrors('startsWithA', 'lengthGreaterThan2'),
    ).not.toEqual(null);

    validationResult = useValidators('Beth', {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(validationResult.isValid).toEqual(false);
    expect(validationResult.validationErrors).toEqual(['startsWithA']);
    expect(
      validationResult.showValidationErrors('startsWithA'),
    ).not.toEqual(null);
    expect(
      validationResult.showValidationErrors('lengthGreaterThan2'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors('startsWithA', 'lengthGreaterThan2'),
    ).not.toEqual(null);

    validationResult = useValidators('B', {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(validationResult.isValid).toEqual(false);
    expect(validationResult.validationErrors).toEqual(
      expect.arrayContaining(['startsWithA', 'lengthGreaterThan2']),
    );
    expect(
      validationResult.showValidationErrors('startsWithA'),
    ).not.toEqual(null);
    expect(
      validationResult.showValidationErrors('lengthGreaterThan2'),
    ).not.toEqual(null);
    expect(
      validationResult.showValidationErrors('startsWithA', 'lengthGreaterThan2'),
    ).not.toEqual(null);

    validationResult = useValidators('Ants', {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(validationResult.isValid).toEqual(true);
    expect(validationResult.validationErrors).toEqual([]);
    expect(
      validationResult.showValidationErrors('startsWithA'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors('lengthGreaterThan2'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors('startsWithA', 'lengthGreaterThan2'),
    ).toEqual(null);
  });

  it('isValid should be true iff all validators predicates return true for each item in list', () => {
    let validationResult = useListValidators(['An', 'Ants'], {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(validationResult.isValid).toEqual(false);
    expect(validationResult.validationErrors).toEqual([['lengthGreaterThan2'], []]);
    expect(
      validationResult.showValidationErrors(0, 'startsWithA'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(0, 'lengthGreaterThan2'),
    ).not.toEqual(null);
    expect(
      validationResult.showValidationErrors(0, 'startsWithA', 'lengthGreaterThan2'),
    ).not.toEqual(null);

    expect(
      validationResult.showValidationErrors(1, 'startsWithA'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(1, 'lengthGreaterThan2'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(1, 'startsWithA', 'lengthGreaterThan2'),
    ).toEqual(null);

    validationResult = useListValidators(['Beth', 'Ants'], {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(validationResult.isValid).toEqual(false);
    expect(validationResult.validationErrors).toEqual([['startsWithA'], []]);
    expect(
      validationResult.showValidationErrors(0, 'startsWithA'),
    ).not.toEqual(null);
    expect(
      validationResult.showValidationErrors(0, 'lengthGreaterThan2'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(0, 'startsWithA', 'lengthGreaterThan2'),
    ).not.toEqual(null);

    expect(
      validationResult.showValidationErrors(1, 'startsWithA'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(1, 'lengthGreaterThan2'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(1, 'startsWithA', 'lengthGreaterThan2'),
    ).toEqual(null);

    validationResult = useListValidators(['B', 'Ants'], {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(validationResult.isValid).toEqual(false);
    expect(validationResult.validationErrors).toEqual(
      [expect.arrayContaining(['startsWithA', 'lengthGreaterThan2']), []],
    );
    expect(
      validationResult.showValidationErrors(0, 'startsWithA'),
    ).not.toEqual(null);
    expect(
      validationResult.showValidationErrors(0, 'lengthGreaterThan2'),
    ).not.toEqual(null);
    expect(
      validationResult.showValidationErrors(0, 'startsWithA', 'lengthGreaterThan2'),
    ).not.toEqual(null);

    expect(
      validationResult.showValidationErrors(1, 'startsWithA'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(1, 'lengthGreaterThan2'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(1, 'startsWithA', 'lengthGreaterThan2'),
    ).toEqual(null);

    validationResult = useListValidators(['Ants', 'Ants'], {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(validationResult.isValid).toEqual(true);
    expect(validationResult.validationErrors).toEqual([[], []]);
    expect(
      validationResult.showValidationErrors(0, 'startsWithA'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(0, 'lengthGreaterThan2'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(0, 'startsWithA', 'lengthGreaterThan2'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(1, 'startsWithA'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(1, 'lengthGreaterThan2'),
    ).toEqual(null);
    expect(
      validationResult.showValidationErrors(1, 'startsWithA', 'lengthGreaterThan2'),
    ).toEqual(null);
  });

  it('show validation errors should render correctly', () => {
    const validationResult = useValidators('B', {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(
      validationResult.showValidationErrors('lengthGreaterThan2'),
    ).toMatchSnapshot();
    expect(
      validationResult.showValidationErrors('startsWithA'),
    ).toMatchSnapshot();
    expect(
      validationResult.showValidationErrors('startsWithA', 'lengthGreaterThan2'),
    ).toMatchSnapshot();
  });

  it('show list validation errors should render correctly', () => {
    const validationResult = useListValidators(['B'], {
      lengthGreaterThan2,
      startsWithA,
    });
    expect(
      validationResult.showValidationErrors(0, 'lengthGreaterThan2'),
    ).toMatchSnapshot();
    expect(
      validationResult.showValidationErrors(0, 'startsWithA'),
    ).toMatchSnapshot();
    expect(
      validationResult.showValidationErrors(0, 'startsWithA', 'lengthGreaterThan2'),
    ).toMatchSnapshot();
  });
});
