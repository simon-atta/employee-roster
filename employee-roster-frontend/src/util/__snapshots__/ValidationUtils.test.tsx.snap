// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation Utils show list validation errors should render correctly 1`] = `
<InputGroupText>
  <Tooltip
    content={
      <List
        style={
          Object {
            "color": "white",
          }
        }
      >
        <ListItem>
          "B" should have a length of at least 3
        </ListItem>
      </List>
    }
  >
    <WarningTriangleIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  </Tooltip>
</InputGroupText>
`;

exports[`Validation Utils show list validation errors should render correctly 2`] = `
<InputGroupText>
  <Tooltip
    content={
      <List
        style={
          Object {
            "color": "white",
          }
        }
      >
        <ListItem>
          "B" should start with "A"
        </ListItem>
      </List>
    }
  >
    <WarningTriangleIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  </Tooltip>
</InputGroupText>
`;

exports[`Validation Utils show list validation errors should render correctly 3`] = `
<InputGroupText>
  <Tooltip
    content={
      <List
        style={
          Object {
            "color": "white",
          }
        }
      >
        <ListItem>
          "B" should start with "A"
        </ListItem>
        <ListItem>
          "B" should have a length of at least 3
        </ListItem>
      </List>
    }
  >
    <WarningTriangleIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  </Tooltip>
</InputGroupText>
`;

exports[`Validation Utils show validation errors should render correctly 1`] = `
<InputGroupText>
  <Tooltip
    content={
      <List
        style={
          Object {
            "color": "white",
          }
        }
      >
        <ListItem>
          "B" should have a length of at least 3
        </ListItem>
      </List>
    }
  >
    <WarningTriangleIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  </Tooltip>
</InputGroupText>
`;

exports[`Validation Utils show validation errors should render correctly 2`] = `
<InputGroupText>
  <Tooltip
    content={
      <List
        style={
          Object {
            "color": "white",
          }
        }
      >
        <ListItem>
          "B" should start with "A"
        </ListItem>
      </List>
    }
  >
    <WarningTriangleIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  </Tooltip>
</InputGroupText>
`;

exports[`Validation Utils show validation errors should render correctly 3`] = `
<InputGroupText>
  <Tooltip
    content={
      <List
        style={
          Object {
            "color": "white",
          }
        }
      >
        <ListItem>
          "B" should start with "A"
        </ListItem>
        <ListItem>
          "B" should have a length of at least 3
        </ListItem>
      </List>
    }
  >
    <WarningTriangleIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  </Tooltip>
</InputGroupText>
`;
